### PR TITLE
fix: Another attempt fixing nested fenced divs keeping structure

### PR DESCRIPTION
**Pull request by Omikhleia**
PR #51 (for issue #44) got merged a bit to quickly to my taste. Here is a PR rebased on it, but reverting to the previous writer API and trying to keep the hierarchical structure. It passes the tests for #51. I wish I had more time to check for edge cases, but at least you have it here for scrutiny.

### CODE REVIEW

File: lunamark/reader/markdown.lua [hunk 1]
```diff
@@ -1312,27 +1290,14 @@ function M.new(writer, options)
...
-                             if div_level <= 0 then
-                               result = nil
-                             else
-                               div_level = div_level - 1
-                               result = i
-                             end
-                             return result
-                           end)
+    larsers.FencedDiv = P{ (larsers.fenced_div_begin)
+                           * C(((parsers.any - (larsers.fenced_div_begin + larsers.fenced_div_end)) + V(1))^0)
```
Comment (Witiko) [inline lunamark/reader/markdown.lua:1294]: How does this grammar handle edge cases such as the following:

````` md
::: {.some-classname}
This should not parse as a div.
``` md
:::
```
`````

File: lunamark/reader/markdown.lua [hunk 2]
```diff
@@ -1312,27 +1290,14 @@ function M.new(writer, options)
...
+                           * larsers.fenced_div_end } / function (attr, div)
+                                                          -- The \n here are to force paragraphs in the ..div
+                                                          -- Not sure Pandoc would generate Para nodes, rather than
+                                                          -- Plain, though!
+                                                          return parse_blocks("\n"..div.."\n"), attr
```
Comment (Witiko) [inline lunamark/reader/markdown.lua:1299]: It would. That's what https://github.com/jgm/lunamark/pull/51/commits/6dbd9244130e58e90de4596fda4ed57cf895ddfa in #51 was for.

**Review by Witiko — CHANGES_REQUESTED** [review]
We may not need to use the full markdown grammar while looking for the end of a div, but it seems to me that we need to consider at least code fences.

File: lunamark/reader/markdown.lua (hunk 1 above)
Comment (Omikhleia) [reply lunamark/reader/markdown.lua:1294]: It breaks here, yes. The edge-case is dubious (how likely people are going to do this?), but granted, it's not what we would expect. As said, I have lacked time to check edge cases. And I do lack time to advance as fast you wanted, so I pushed what I had started.

It spits out:
~~~
<div class="some-classname">This should not parse as a div.

<p>``` md</p></div>

<p>```</p>
~~~

This is bad. **But** your own version now on master, unless mistaken, spits out:
~~~
<div class="some-classname">

This should not parse as a div.

<pre><code class="language-md">:::
</code></pre>
~~~

Which is not better (and even lacks the closing div?). Or did I miss something?

Comment (Omikhleia) [reply lunamark/reader/markdown.lua:1294]: Wouldn't something like this solve the fenced code block case:
```
    larsers.fencedcodeblock = parsers.TildeFencedCodeBlock + parsers.BacktickFencedCodeBlock
    larsers.FencedDiv = P{ (larsers.fenced_div_begin)
                           * C((((parsers.any - larsers.fencedcodeblock - (larsers.fenced_div_begin + larsers.fenced_div_end))) + V(1) + larsers.fencedcodeblock)^0)
                           * larsers.fenced_div_end } / function (attr, div)
                                                          -- The \n here are to force paragraphs in the ..div
                                                          -- Not sure Pandoc would generate Para nodes, rather than
                                                          -- Plain, though!
                                                          return parse_blocks("\n"..div.."\n"), attr
                                                        end / writer.div 
                                                        ...
```
(without considering there to cancel `larsers.fencedcodeblock` if the corresponding option is not enabled)
Or am I totally off ! (Which is possible too... LPeg is hard to grasp sometimes!)

Comment (Witiko) [reply lunamark/reader/markdown.lua:1294]: I think you are on the right track.

**Comment by Witiko** [comment]
Closes https://github.com/jgm/lunamark/issues/53.

File: lunamark/reader/markdown.lua (hunk 1 above)
Comment (Witiko) [reply lunamark/reader/markdown.lua:1294]: > Or did I miss something?

Can you please add https://github.com/jgm/lunamark/pull/52#discussion_r1030976336 and https://github.com/jgm/lunamark/issues/53 to the unit tests, so that we can see how the code is failing at the moment and fix it? I don't have a local environment,where I could test lunamark.

**Comment by Witiko** [comment]
This issue should also close https://github.com/jgm/lunamark/issues/53 and https://github.com/jgm/lunamark/issues/54. @Omikhleia, can you please add https://github.com/jgm/lunamark/pull/52#discussion_r1030976336, https://github.com/jgm/lunamark/issues/53, https://github.com/jgm/lunamark/issues/54 to the unit tests, so that we can at least see the extent to which the code is failing at the moment?

> PR https://github.com/jgm/lunamark/pull/51 (for issue https://github.com/jgm/lunamark/issues/44) got merged a bit to quickly to my taste.

I am sorry about that. I should have heeded your concerns. I am way behind on my schedule for the monthly release of witiko/markdown due to the bugged fenced div implementation, but rushing ahead and introducing new bugs to `master` seems reckless (although we are moving forward!).

**Comment by Omikhleia** [comment]
@Witiko 
Div-blockquote tests we might want (if not already covered)
~~~
Here is a properly nested div in a block quote:

::: {.level1}
This is the beginning of a div

> This is a blockquote
>
> :::: {.level2-inside-bloquote}
> This is a inside inside a block quote
> ::::
:::

As nested, the quoted colons do not close the embedding div.

:::::: {.some-classname}
This is the beginning of a div

> This is a blockquote
> ::::::

::::::
~~~

The first is a "regular" case that works, the second currently breaks.

> Will you find the time to tackle this by the end of the month, or should I?

I won't have the bandwidth to guarantee I can work on it next week (a probably a bit more). Feel free to go ahead. I'll try to follow the PR though, for comments. (And if at some point you want to squash commits, no problem either. A `Co-authored-by: name <name@example.com>`  comment might then be useful - but not mandated in any way, I just discovered that GitHub trick recently).

**Comment by Witiko** [comment]
> The first is a "regular" case that works, the second currently breaks.

Good thinking. I will add the regular case.

> (And if at some point you want to squash commits, no problem either. A Co-authored-by: name <name@example.com> comment might then be useful - but not mandated in any way, I just discovered that GitHub trick recently).

Thanks, I did not know that. I will also do that for commits based on your comments.

**Comment by Witiko** [comment]
> I won't have the bandwidth to guarantee I can work on it next week (and probably a bit more).

I have postponed the 2.19.0 release of witiko/markdown to December 23, so this is less urgent for me. Besides fenced divs, I will also need https://github.com/jgm/lunamark/issues/43 closed before the release. If you'd like, we can split the effort and I will focus on https://github.com/jgm/lunamark/issues/43.

If you'd like to have a direct channel to discuss the implementation, please feel free to join the witiko/markdown [discord server][1] or [matrix space][2].

 [1]: https://discord.gg/8xJsPghzSH
 [2]: https://matrix.to/#/#witiko-markdown:matrix.org

**Comment by Witiko** [comment]
@Omikhleia I have fixed all the test cases except for two related to fenced code blocks:
https://github.com/jgm/lunamark/blob/2d70835c8e56366ae94eb4575612da472a35f94f/tests/lunamark/ext_fenced_divs.test#L66-L78

A big issue was correctly handling newlines: we want newlines before `larsers.fenced_code_begin` and `larsers.fenced_code_end` but only if there is some intervening text in between. In the end, I grabbed a piece of paper and drew the following automaton:

state-machine

Here is the corresponding LPEG pattern:

https://github.com/jgm/lunamark/blob/2d70835c8e56366ae94eb4575612da472a35f94f/lunamark/reader/markdown.lua#L1308-L1321

Seems to work like charm.

**Comment by Witiko** [comment]
@Omikhleia It appears that even if we solve fenced divs, our approach is still incompatible with Pandoc. Let's take the following example:

``` md
::: {.not-a-div}
- here is a list item
:::
that stretches across several lines
```

For this example, Pandoc will produce the following HTML output:

``` html
<p>::: xxx - here is a list item ::: that that stretches across several lines</p>
```

I added this example to the unit tests.

What this tells me is that Pandoc doesn't do simplistic forward-scanning like we do. Instead, it parses blocks using the full grammar and looks for the closing tag (`:::`) only at block boundaries. We could do the same by folding `larsers.FencedDiv` into the full grammar by removing `P{ ... }` and by replacing `V(1)` with `V("Blocks")` and adding and exception for `larsers.fenced_div_end` to `larsers.Endline` and `larsers.NonbreakingEndline`.

The advantage of this change is that we don't need to write a special forward-scanner for fenced divs. Instead, we just use the full grammar of markdown. Arguably, this will be significantly easier to maintain going forward.

**Comment by Witiko** [comment]
@Omikhleia I am done and would appreciate your review.

> - adding and exception for `larsers.fenced_div_end` to `larsers.Endline` and `larsers.NonbreakingEndline`.

This ended up being trickier than I expected. Assume the following input:

``` md
:::
This is not a div
:::
```

If we add an exception for `larsers.fenced_div_end` to `larsers.Endline` and `larsers.NonbreakingEndline`, then this will produce the following output:

``` html
<p>::: this is not a div</p>
<p>:::</p>
```

Therefore, we are keeping the nesting level as a named capture inside the grammar and we only match the exception for `larsers.fenced_div_end` if we are nested inside a div:

https://github.com/jgm/lunamark/blob/d247439f38fcde5194637410d43b5a5c4e4b0b22/lunamark/reader/markdown.lua#L1715-L1718
https://github.com/jgm/lunamark/blob/d247439f38fcde5194637410d43b5a5c4e4b0b22/lunamark/reader/markdown.lua#L1322-L1343
https://github.com/jgm/lunamark/blob/d247439f38fcde5194637410d43b5a5c4e4b0b22/lunamark/reader/markdown.lua#L1124-L1136

This is kind of clever and [follows a similar trick done by Pandoc][1]. However, I welcome any ideas for simplification.

 [1]: https://github.com/jgm/pandoc/blob/ed4bdd059cc6c4b8045b2c4c2c0f8b44d3faa3d6/src/Text/Pandoc/Readers/Markdown.hs#L2085-L2089

**Comment by Witiko** [comment]
One last quibble I have with the current implementation is that a div must end with a blank line, otherwise it will not be enclosed in a paragraph. Let's take the following example:

``` md
::: {.myclass}
Some div
:::
```

Currently, Lunamark will produce the following output:
``` html
<div class="myclass">Some div</div>
```

Your original solution from this PR called `parse_blocks`, where we could artificially insert newlines at the end of the div. I am not sure we can easily do that with my solution. However, this trouble with paragraphs seems to affect lunamark in general:

```
$ printf '> xxx' | lunamark
<blockquote>
xxx
</blockquote>

$ printf '> xxx\n' | lunamark
<blockquote>
<p>xxx</p>
</blockquote>
```

Regardless of whether this is a feature or a bug, it seems easily solved by setting `writer.paragraph = writer.plain`.

**Comment by Omikhleia** [comment]
@Witiko What an amazing work you have done here! I'm sorry I couldn't contribute - Not only did I expected the week to be very busy, but I am sick with my brain at 10% capacity... Just took the time to test you branch on a bunch of small test I had (mostly redundant with the ones you added, but written a bit differently, so who knows...) = They all went right...

There's _one_ case where I didn't get same result as Pandoc and was a bit astonished at first:
```
::::: {.some-classname}
This is the beginning of a div

> This is a blockquote
> :::::
> but not a nested div

:::
```

Pandoc gives
```
<div class="some-classname">
<p>This is the beginning of a div</p>
<blockquote>
<p>This is a blockquote</p>
<p>::::: but not a nested div</p>
</blockquote>
</div>
```
Lunamark with your fixes (some line breaks added for easier comparison)
```
<div class="some-classname">
<p>This is the beginning of a div</p>
<blockquote>
<p>This is a blockquote ::::: but not a nested div</p>
</blockquote>
</div>
```

Small discrepancy for a "broken" case - and I am even thinking our output might be more correct maybe.

I'll look at code as soon as I can and have the necessary focus^^

**Comment by Witiko** [comment]
> Small discrepancy for a "broken" case - and I am even thinking our output might be more correct maybe.

Definitely looks like a bug. I will test this with up-to-date Pandoc and optionally report it.

> I am sick with my brain at 10% capacity... I'll look at code as soon as I can and have the necessary focus^^

I am sorry to hear that! Please, take all the time you need.

**Comment by Omikhleia** [comment]
> but I haven't been able to pinpoint it this morning.

Doh, it was in front of my eyes: `larsers.Endline` and `larsers.NonbreakingEndline` both use `larsers.fencestart`

**Comment by Witiko** [comment]
This seems as a somewhat random thing to do: we definitely don't want to reset the div_level to 0 at inlines. Instead we should investigate why div_level is not defined here.

**Comment by Witiko** [comment]
Seems reasonable. I just hope there do not exist other copies of `syntax` that change the root rule and may need similar treatment.

**Review by Witiko — APPROVED** [review]

**Comment by Omikhleia** [comment]
LGTM !

I just successfully tested two cases I had seen where end-of-lines were breaking. The inline notes, as noted above:
```
Note^[Some
note]
```
but also the indirect links, like the following:
```
Link
[some
link](www.google.com "toto titi")
```
So besides `larsers.inlines`, this covers the uses of `inlines_no_inline_note` and `inlines_no_link_t`. The result is now the same, whether `fenced_divs` is enable or not.

Only remaining item would be `inlines_nbsp` but I have no idea how to check it - it's used in citations, and I never used them so far (just browsed about them and tried, but I don't how to trigger the case in question). This being said, I guess we could assume being covered by the two other cases, since it's (more or less) the same logic. So from my viewpoint, it's all good.

File: lunamark/reader/markdown.lua [hunk 3]
```diff
@@ -1325,27 +1319,28 @@ function M.new(writer, options)
...
-                               result = i
-                             end
-                             return result
-                           end)
+    return Cg( Cmt(Cb("div_level"), update_div_level)
+             , "div_level")
+  end
+
+  larsers.FencedDiv = larsers.fenced_div_begin * increment_div_level(1)
+                    * parsers.blanklines
```
Comment (Omikhleia) [inline lunamark/reader/markdown.lua:1334]:
```suggestion
                    * parsers.skipblanklines
```

Seems this solves the problem with
```
::: { .cit }

I am a _fenced_ div
:::

Rendering as an empty div

File: lunamark/reader/markdown.lua [hunk 4]
```diff
@@ -1325,27 +1319,28 @@ function M.new(writer, options)
...
+                    * Ct( (V("Block") - larsers.fenced_div_end)^-1
+                        * (parsers.blanklines / function()
+                                              return writer.interblocksep
+                                            end
+                          * (V("Block") - larsers.fenced_div_end))^0)
+                    * parsers.blanklines
```
Comment (Omikhleia) [inline lunamark/reader/markdown.lua:1340]:
```suggestion
                    * parsers.skipblanklines
```
Perhaps here too?

File: lunamark/reader/markdown.lua (hunk 3 above)
Comment (Omikhleia) [reply lunamark/reader/markdown.lua:1334]: Moving `parsers.blanklines` in the Ct capture would work too, but the output is less convincing (we get a blank line).

**Comment by Witiko** [comment]
The current implementation seems quite robust. Hopefully, we have not overlooked anything major.
